### PR TITLE
Add class-wise metrics logging and confusion matrix to DeepMIL

### PR DESCRIPTION
**Pull request by harshita-s**
In this PR:
- Absolute confusion matrix for test set is printed along with other metrics at the end of the run
- Normalized confusion matrix for test set is saved as a heatmap figure in outputs/fig
- Per-class accuracy is logged every epoch for train and validation stages (plots are visible along with other metrics)
- Class names are added as an argument to DeepMIL (could be passed through the containers depending on task, default None)
- Test for normalization and plotting the normalized confusion matrix

### CODE REVIEW

File: InnerEye/ML/Histopathology/models/deepmil.py [hunk 1]
```diff
@@ -53,7 +55,8 @@ def __init__(self,
                  verbose: bool = False,
                  slide_dataset: SlidesDataset = None,
                  tile_size: int = 224,
-                 level: int = 1) -> None:
+                 level: int = 1,
+                 class_names: List[str] = None) -> None:
```
Comment (ant0nsc) [inline InnerEye/ML/Histopathology/models/deepmil.py:59]: Type annotation has a small discrepancy - if you set the default to None, it should be "Optional[List[str]]".

File: InnerEye/ML/Histopathology/models/deepmil.py [hunk 2]
```diff
@@ -84,6 +87,14 @@ def __init__(self,
         self.encoder = encoder
         self.num_encoding = self.encoder.num_encoding
 
+        if class_names is not None:
+            self.class_names = class_names
```
Comment (ant0nsc) [inline InnerEye/ML/Histopathology/models/deepmil.py:91]: you should probably check if the class names and n_classes are in sync.

File: InnerEye/ML/Histopathology/models/deepmil.py [hunk 3]
```diff
@@ -148,21 +159,29 @@ def get_metrics(self) -> nn.ModuleDict:
...
             return nn.ModuleDict({'accuracy': Accuracy(),
-                                   'auroc': AUROC(num_classes=self.n_classes),
-                                   'precision': Precision(),
-                                   'recall': Recall(),
-                                   'f1score': F1()})
+                                  'auroc': AUROC(num_classes=self.n_classes),
+                                  'precision': Precision(),
+                                  'recall': Recall(),
+                                  'f1score': F1(),
+                                  'confusion_matrix': ConfusionMatrix(num_classes=self.n_classes+1)})
```
Comment (ant0nsc) [inline InnerEye/ML/Histopathology/models/deepmil.py:170]: Seeing this here: It would be good to add documentation for n_classes beyond what you have now. For two classes "0" and "1", n_classes should be set to 1, correct?

File: InnerEye/ML/Histopathology/models/deepmil.py [hunk 4]
```diff
@@ -148,21 +159,29 @@ def get_metrics(self) -> nn.ModuleDict:
...
 
     def log_metrics(self,
                     stage: str) -> None:
         valid_stages = ['train', 'test', 'val']
         if stage not in valid_stages:
             raise Exception(f"Invalid stage. Chose one of {valid_stages}")
         for metric_name, metric_object in self.get_metrics_dict(stage).items():
-            self.log(f'{stage}/{metric_name}', metric_object, on_epoch=True, on_step=False, logger=True, sync_dist=True)
+            if not metric_name == "confusion_matrix":
+                self.log(f'{stage}/{metric_name}', metric_object, on_epoch=True, on_step=False, logger=True, sync_dist=True)
```
Comment (ant0nsc) [inline InnerEye/ML/Histopathology/models/deepmil.py:179]: can you replace the self.log calls with hi-ml's "log_on_epoch" function? that gives you a simpler interface and handle sync_dist better.

Comment (ant0nsc) [reply InnerEye/ML/Histopathology/models/deepmil.py:179]: (also in the else branch)

File: InnerEye/ML/Histopathology/models/deepmil.py [hunk 5]
```diff
@@ -148,21 +159,29 @@ def get_metrics(self) -> nn.ModuleDict:
...
+                                  'confusion_matrix': ConfusionMatrix(num_classes=self.n_classes+1)})
 
     def log_metrics(self,
                     stage: str) -> None:
         valid_stages = ['train', 'test', 'val']
         if stage not in valid_stages:
             raise Exception(f"Invalid stage. Chose one of {valid_stages}")
         for metric_name, metric_object in self.get_metrics_dict(stage).items():
-            self.log(f'{stage}/{metric_name}', metric_object, on_epoch=True, on_step=False, logger=True, sync_dist=True)
+            if not metric_name == "confusion_matrix":
```
Comment (ant0nsc) [inline InnerEye/ML/Histopathology/models/deepmil.py:178]: I see you have some `if` statements that are "if not A then something() else otherthing()". If you have anyway handling both cases, it is easier to read the code if you do not negate the condition and swap if/else.

File: InnerEye/ML/Histopathology/models/deepmil.py [hunk 6]
```diff
@@ -338,6 +357,18 @@ def test_epoch_end(self, outputs: List[Dict[str, Any]]) -> None:  # type: ignore
         fig = plot_scores_hist(results)
         self.save_figure(fig=fig, figpath=outputs_fig_path / 'hist_scores.png')
 
+        print("Computing and saving confusion matrix...")
+        metrics_dict = self.get_metrics_dict('test')
+        cf_matrix = metrics_dict["confusion_matrix"].compute()
```
Comment (ant0nsc) [inline InnerEye/ML/Histopathology/models/deepmil.py:362]: This literal "confusion_matrix" needs to be in sync with your other uses of "confusion_matrix". This is an extremely common source of errors - you change the constant somewhere, and forget to change is somewhere else (as trivial/benign as this may sound). Your code will be a lot safer (and require fewer tests) if you define those literals as constants, `CONF_MATRIX_METRIC = "confusion_matrix"` and re-use it.

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILECrck.py [hunk 7]
```diff
@@ -135,12 +135,23 @@ def get_path_to_best_checkpoint(self) -> Path:
...
+                                        self.checkpoint_folder_path,
+                                        self.best_checkpoint_filename_with_suffix)
+        if absolute_checkpoint_path.is_file():
+            return absolute_checkpoint_path
+
+        absolute_checkpoint_path_parent = Path(fixed_paths.repository_parent_directory(),
+                                    self.checkpoint_folder_path,
+                                    self.best_checkpoint_filename_with_suffix)
+        if absolute_checkpoint_path_parent.is_file():
+            return absolute_checkpoint_path_parent
```
Comment (ant0nsc) [inline InnerEye/ML/configs/histo_configs/classification/DeepSMILECrck.py:148]: Confused. This and the variable above are exactly the same?

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILECrck.py [hunk 8]
```diff
@@ -135,12 +135,23 @@ def get_path_to_best_checkpoint(self) -> Path:
...
-            fixed_paths.repository_root_directory()
-            / self.checkpoint_folder_path
-            / self.best_checkpoint_filename_with_suffix
-        )
-
+        absolute_checkpoint_path = Path(fixed_paths.repository_root_directory(),
+                                        self.checkpoint_folder_path,
+                                        self.best_checkpoint_filename_with_suffix)
+        if absolute_checkpoint_path.is_file():
+            return absolute_checkpoint_path
```
Comment (ant0nsc) [inline InnerEye/ML/configs/histo_configs/classification/DeepSMILECrck.py:142]: I'm puzzled. Checkpoints are normally stored in the "outputs" folder, so that they are also available at the end of an AzureML run. Why are the checkpoints here accessed as part of repository root?

File: Tests/ML/histopathology/utils/test_metrics_utils.py [hunk 9]
```diff
@@ -126,6 +126,30 @@ def test_plot_heatmap_overlay(test_output_dirs: OutputFolderForTests) -> None:
...
+    assert (cm_n <= 1).all()
+
+    fig = plot_normalized_confusion_matrix(cm=cm_n, class_names=class_names)
+    assert isinstance(fig, matplotlib.figure.Figure)
+    file = Path(test_output_dirs.root_dir) / f"plot_confusion_matrix_{n_classes}.png"
+    resize_and_save(5, 5, file)
+    assert file.exists()
+    expected = full_ml_test_data_path("histo_heatmaps") / f"confusion_matrix_{n_classes}.png"
+    # To update the stored results, uncomment this line:
+    expected.write_bytes(file.read_bytes())
```
Comment (ant0nsc) [inline Tests/ML/histopathology/utils/test_metrics_utils.py:149]: This should not be a checked in - your test will always pass now!
```suggestion
    # expected.write_bytes(file.read_bytes())
```

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILECrck.py (hunk 7 above)
Comment (harshita-s) [reply InnerEye/ML/configs/histo_configs/classification/DeepSMILECrck.py:148]: I have taken this from our other config. https://github.com/microsoft/InnerEye-DeepLearning/blob/fb258d5c9a3ba18565b5a67e7ac1f00127d9ecb9/InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py#L148 It was only to enable this config to work and may not be related to this PR. Will remove for now for another PR

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILECrck.py (hunk 8 above)
Comment (harshita-s) [reply InnerEye/ML/configs/histo_configs/classification/DeepSMILECrck.py:142]: I have taken this from our other config https://github.com/microsoft/InnerEye-DeepLearning/blob/fb258d5c9a3ba18565b5a67e7ac1f00127d9ecb9/InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py#L148
It was only to enable the Crck config to work and may not be related to this PR. Will remove for now for another PR

File: InnerEye/ML/Histopathology/models/deepmil.py (hunk 2 above)
Comment (harshita-s) [reply InnerEye/ML/Histopathology/models/deepmil.py:89]: Thanks, I have now added a check

File: InnerEye/ML/Histopathology/models/deepmil.py (hunk 1 above)
Comment (harshita-s) [reply InnerEye/ML/Histopathology/models/deepmil.py]: Thanks, now changed

File: InnerEye/ML/Histopathology/models/deepmil.py (hunk 3 above)
Comment (harshita-s) [reply InnerEye/ML/Histopathology/models/deepmil.py]: Thanks, added to the docstring

File: InnerEye/ML/Histopathology/models/deepmil.py (hunk 4 above)
Comment (harshita-s) [reply InnerEye/ML/Histopathology/models/deepmil.py]: wondering if log_on_epoch will work with torch module metrics objects?

File: InnerEye/ML/Histopathology/models/deepmil.py (hunk 5 above)
Comment (harshita-s) [reply InnerEye/ML/Histopathology/models/deepmil.py]: Thanks, I swapped if/else

File: InnerEye/ML/Histopathology/models/deepmil.py (hunk 6 above)
Comment (harshita-s) [reply InnerEye/ML/Histopathology/models/deepmil.py]: Thanks for the suggestion, now defined metrics names as constants

File: Tests/ML/histopathology/utils/test_metrics_utils.py (hunk 9 above)
Comment (harshita-s) [reply Tests/ML/histopathology/utils/test_metrics_utils.py]: Thanks for spotting this, changed now

File: InnerEye/ML/Histopathology/models/deepmil.py (hunk 4 above)
Comment (harshita-s) [reply InnerEye/ML/Histopathology/models/deepmil.py]: This is now changed, thanks

**Review by ant0nsc — APPROVED** [review]

**Review by dccastro — APPROVED** [review]
Great job! No other comments 👍
